### PR TITLE
Neural ODE API + Tutorial

### PR DESCRIPTION
**Pull request by shivance**
# Pull Request Template

Through this Pull Request I'll be contributing the proposed NeuralODE API to deepchem, through [this](https://docs.google.com/document/d/1pwu72pVCHhwO1f58gppAx5xD4NUM3hMo8YrwBaMuW9c/edit?usp=sharing) design document.

#2680

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by shivance** [comment]
Unit Tests are yet to be added !

File: deepchem/models/torch_models/layers.py [hunk 1]
```diff
@@ -23,6 +23,173 @@
   from collections import Sequence as SequenceCollection
 
 
+class LinearModule(nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:26]: We should double check there isn't already a linear layer.

Need to add docs and should add to the layers.rst file

File: deepchem/models/torch_models/layers.py [hunk 2]
```diff
@@ -23,6 +23,173 @@
...
+               dropouts: OneOrMany[float] = 0.5,
+               activation_fns: OneOrMany[ActivationFn] = 'relu',
+               weight_init_stddev: OneOrMany[float] = 0.02,
+               bias_init_consts: OneOrMany[float] = 1.0,
+               residual: bool = True,
+               ode_block: nn.Module = None,
+               enable_ffn: bool = True):
+    """Linear Feed Forward Network
+
+      TODO : Flexibility to use Convolutional/ Linear NN as modelling network
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:48]: Some formatting issues here

File: deepchem/models/torch_models/layers.py [hunk 3]
```diff
@@ -23,6 +23,173 @@
...
+               n_tasks: int = 1,
+               n_features: int = 1,
+               n_classes: int = 1,
+               mode: str = "regression",
+               dropouts: OneOrMany[float] = 0.5,
+               activation_fns: OneOrMany[ActivationFn] = 'relu',
+               weight_init_stddev: OneOrMany[float] = 0.02,
+               bias_init_consts: OneOrMany[float] = 1.0,
+               residual: bool = True,
+               ode_block: nn.Module = None,
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:44]: A linear module is general, and shouldn't know about `ode_block`

File: deepchem/models/torch_models/layers.py [hunk 4]
```diff
@@ -23,6 +23,173 @@
...
+
+    if self.ode_block is not None:
+      x, _ = self.ode_block(x)
+
+    outputs = x
+
+    if self.enable_ffn:
+      batch_size = x.shape[0]
+
+      if self.mode == "classification":
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:177]: A linear layer shouldn't know about classification/regression

File: deepchem/models/torch_models/neuralode.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+  datasets
+  
+  TODO : Add Usage Example
+
+  """
+
+  def __init__(self,
+               n_tasks: int,
+               n_features: int,
+               vector_field_config: dict,
```
Comment (rbharath) [inline deepchem/models/torch_models/neuralode.py:35]: Have explicit flags instead of configs

File: deepchem/models/torch_models/neuralode.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+           ids_b) in dataset.iterbatches(batch_size=self.batch_size,
+                                         deterministic=deterministic,
+                                         pad_batches=pad_batches):
+        if mode == 'predict':
+          dropout = np.array(0.0)
+        else:
+          dropout = np.array(1.0)
+        yield ([t_b, dropout], Y_b, [w_b])
+
+  # TODO : Write custom fit method
```
Comment (rbharath) [inline deepchem/models/torch_models/neuralode.py:216]: We probably shouldn't have a custom fit method. Try double checking that the regular fit won't work

File: deepchem/models/torch_models/neuralode.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+  allows users to
+  
+  - Configurable the neural net modelling system dynamics in terms of number of layers,
+  dropouts, weight initialisation
+  - Configurable ode solver, allowing to set various sensitivity, adjoint methods for
+  ode solvers
+  - Custom Wrapper NN to NeuralODE enabling to use for deepchem tasks, and fit on deepchem 
+  datasets
+  
+  TODO : Add Usage Example
```
Comment (rbharath) [inline deepchem/models/torch_models/neuralode.py:28]: Once we have unit tests and usage examples, we can see whether the API is clean and works

File: deepchem/models/torch_models/neuralode.py (hunk 5 above)
Comment (shivance) [reply deepchem/models/torch_models/neuralode.py:35]: Done !

File: deepchem/models/torch_models/layers.py (hunk 4 above)
Comment (shivance) [reply deepchem/models/torch_models/layers.py:177]: encapsulating the vector field layer within neuralode model
[reference](https://github.com/deepchem/deepchem/blob/master/deepchem/models/fcnet.py)

**Pull request closed without merging**